Add optional fixes for loading skins in legacy versions

TheKodeToad opened the pull request:
This is heavily inspired by [this PR by craftycodie](https://github.com/Mojang/LegacyLauncher/pull/33).

Changes:
- Add online fixes, currently just legacy skin fix.
- Move all legacy classes to a separate JAR. This is done in a slightly hacky way, but there doesn't seem to be an easy way to separately compile and zip Java classes with CMake.

Potential future changes:
- Online *mode* fix
- Classic level saving
- Resources fix
- Everything from https://github.com/craftycodie/MineOnline/tree/main/src/main/java/gg/codie/mineonline/protocol
- Window resizing on versions that don't support it (not really applicable for online fixes, just legacy fixes in general)

Depends on https://github.com/PrismLauncher/meta/pull/10 being merged.

![In action](https://user-images.githubusercontent.com/57493648/201476120-9f8dd6de-bd62-4b13-b072-a752271978ae.png)

![image](https://user-images.githubusercontent.com/57493648/209463128-f56fcc47-2b32-42f4-a130-268b116951e0.png)


<!--
Hey there! Thanks for your contribution.

Please make sure that your commits are signed off first.
If you don't know how that works, check out our contribution guidelines: https://github.com/PrismLauncher/PrismLauncher/blob/develop/CONTRIBUTING.md#signing-your-work
If you already created your commits, you can run `git rebase --signoff develop` to retroactively sign-off all your commits and `git push --force` to override what you have pushed already.

Note that signing and signing-off are two different things!
-->

TheKodeToad committed:
Experimental skin fix, inspired by craftycodie's

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Add separate util class

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Only use DatatypeConverter as a fallback

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

DioEgizio commented:
94e672d7d13ac93216d0f2d06ee6ae1bd5f0aa4d why is this necessary?

GitHub Runners already have java (11)

TheKodeToad committed:
Add workaround to warning

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad commented:
> [94e672d](https://github.com/PrismLauncher/PrismLauncher/commit/94e672d7d13ac93216d0f2d06ee6ae1bd5f0aa4d) why is this necessary?
> 
> GitHub Runners already have java (11)

Because Java 8 is required to build it now since it uses DatatypeConverter as a fallback for Base64. DatatypeConverter is not available in 11.

DioEgizio commented:
> > [94e672d](https://github.com/PrismLauncher/PrismLauncher/commit/94e672d7d13ac93216d0f2d06ee6ae1bd5f0aa4d) why is this necessary?
> > GitHub Runners already have java (11)
> 
> Because Java 8 is required to build it now since it uses DatatypeConverter as a fallback for Base64. DatatypeConverter is not available in 11.

Well that's an issue, we still want to support building on newer JDKs

TheKodeToad commented:
> > > [94e672d](https://github.com/PrismLauncher/PrismLauncher/commit/94e672d7d13ac93216d0f2d06ee6ae1bd5f0aa4d) why is this necessary?
> > > GitHub Runners already have java (11)
> > 
> > 
> > Because Java 8 is required to build it now since it uses DatatypeConverter as a fallback for Base64. DatatypeConverter is not available in 11.
> 
> Well that's an issue, we still want to support building on newer JDKs

On Java 7, the minimum version which needs to be supported, Base64 is not available. Java 8 is the only (LTS) version where both are available.

That being said, a workaround with reflection *could* be used. Alternatively, probably the least hacky fix would be just to implement Base64 inside the utility class since I don't think it's that complex.

TheKodeToad commented:
It turns out implementing an efficient base64 decoder is far from being trivial. Who would've thought.
Eventually -source and -target 7 (which is also used) will stop working on the latest javac, so maybe it's best to just accept that Java 10 or earlier will be required to build? Java 8 is still very common.

Scrumplex commented:
needs rebase

TheKodeToad commented:
I've rebased now. I just saw the CodeQL errors, so that's probably another good reason to keep compat with Java 11.

TheKodeToad committed:
Improve the skin fix code
- Spoof 404 instead of keeping original URL.
- Move JsonParseException to the package.
- Pass proxy as null to reduce code duplication.

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Better skin fix error handling

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Use reflection to access DatatypeConverter

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Automatically add add-opens if Java version >= 9

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

craftycodie commented:
@TheKodeToad you could take this opportunity to also add online-mode support to old versions. The implementation is very similar, you can find an example in MineOnline's protocol

https://github.com/craftycodie/MineOnline/tree/main/src/main/java/gg/codie/mineonline/protocol

craftycodie commented on libraries/launcher/org/prismlauncher/fix/skins/SkinFix.java:
```diff
@@ -0,0 +1,91 @@
+    @Override
+    public void apply() {
+    	try {
+    		URL.setURLStreamHandlerFactory(this);
```
You can drop this import by grabbing a reference to the http default handler before replacing the default factory, like so

```java
    private final Class<? extends URLConnection> defaultHttpConnectionClass;

    @Override
    public void apply() {
    	try {
            URL foo = new URL("http://example.com");
            // Doesn't actually establish a connection
            defaultHttpConnectionClass = foo.openConnection().getClass();
            
            URL.setURLStreamHandlerFactory(this);
    	} catch (Error e) {
            Log.warning("Cannot apply skin fix");
    		Log.warning("URLStreamHandlerFactory is already set");
			Log.warning("Turning off legacy skin fix in Settings > Miscellaneous will silence the warnings");
    	}
    }
    	
    @Override
    public URLStreamHandler createURLStreamHandler(String protocol) {
        if ("http".equals(protocol))
            return new Handler(defaultHttpConnectionClass);

        return null;
    }
```

craftycodie replied:
cont.

Then in your handler, like so

```java
        try {
            return defaultHttpConnectionClass.getConstructor(URL.class, Proxy.class).newInstance(url, null);
        } catch (Exception e) {
            // If the constructor isn't found, you can log that out. It's not expected.
            return null;
        }
```

craftycodie reviewed (changes requested):
Nice to see this fix being applied elsewhere, would recommend dropping the nasty `sun.net.www.protocol` entirely as described :)

TheKodeToad commented:
> Nice to see this fix being applied elsewhere, would recommend dropping the nasty `sun.net.www.protocol` entirely as described :)

Hi! It's nice to get a reply from the person who inspired this. Thank you!

I'm a bit confused since it doesn't use the sun classes - it simply obtains the URLStreamHandler instance from a package-private method in java.net.URL.

craftycodie commented:
> > Nice to see this fix being applied elsewhere, would recommend dropping the nasty `sun.net.www.protocol` entirely as described :)
> 
> Hi! It's nice to get a reply from the person who inspired this. Thank you!
> 
> I'm a bit confused since it doesn't use the sun classes - it simply obtains the URLStreamHandler instance from a package-private method in java.net.URL.

Ah! I see! I think i was actually reading the wrong PR, I had both this and my own open 😭 
Apologies, disregard!

TheKodeToad commented:
> > > Nice to see this fix being applied elsewhere, would recommend dropping the nasty `sun.net.www.protocol` entirely as described :)
> > 
> > 
> > Hi! It's nice to get a reply from the person who inspired this. Thank you!
> > I'm a bit confused since it doesn't use the sun classes - it simply obtains the URLStreamHandler instance from a package-private method in java.net.URL.
> 
> Ah! I see! I think i was actually reading the wrong PR, I had both this and my own open sob Apologies, disregard!

Whoops! It's fine, now I understand what happened. I'd still appreciate feedback on my code.

I could try implementing more fixes. Maybe I could make more subclasses of Fix, or just move it all into LegacyOnlineFix or something like that.

craftycodie commented:
> > > > Nice to see this fix being applied elsewhere, would recommend dropping the nasty `sun.net.www.protocol` entirely as described :)
> > > 
> > > 
> > > Hi! It's nice to get a reply from the person who inspired this. Thank you!
> > > I'm a bit confused since it doesn't use the sun classes - it simply obtains the URLStreamHandler instance from a package-private method in java.net.URL.
> > 
> > 
> > Ah! I see! I think i was actually reading the wrong PR, I had both this and my own open sob Apologies, disregard!
> 
> Whoops! It's fine, now I understand what happened. I'd still appreciate feedback on my code.
> 
> I could try implementing more fixes. Maybe I could make more subclasses of Fix, or just move it all into LegacyOnlineFix or something like that.

For sure, probably don't put them all in one PR, makes review/qa more difficult :)

craftycodie commented on launcher/minecraft/MinecraftInstance.cpp:
```diff
@@ -192,6 +192,8 @@ void MinecraftInstance::loadSpecificSettings()
     m_settings->registerSetting("JoinServerOnLaunch", false);
     m_settings->registerSetting("JoinServerOnLaunchAddress", "");
 
+	m_settings->registerSetting("LegacySkinFix", true);
```
nit: inconsistent indenting

TheKodeToad commented:
(I don't think I need to keep quoting the previous message lol)

Yep, that's why I hadn't already. Maybe it is worth getting rid of the Fix interface though since there's probably only going to be one awkward subclass. Separating individual URL fixes probably isn't going to be something users will want, and other fixes will most likely *not* be able to utilise this interface.

craftycodie commented on launcher/minecraft/MinecraftInstance.cpp:
```diff
@@ -438,15 +440,15 @@ QStringList MinecraftInstance::javaArguments()
     args << "-Duser.language=en";
 
+    if (javaVersion.isModular() && traits().contains("legacySkins") && settings()->get("LegacySkinFix").toBool())
+        args << "--add-opens" << "java.base/java.net=ALL-UNNAMED";
```
q: could you explain the purpose of this argument?

TheKodeToad replied:
Without this an error appears on newer Java versions due to using reflection.

craftycodie replied:
Could this not be enabled regardless of the fix then? It sounds like it'd be useful in other cases.
I also think it should be commented to provide context.

TheKodeToad replied:
I agree with adding a comment, but I don't think using this argument is considered good practice. It should only automatically be added if it's necessary, and it also only applies to java.net.

craftycodie commented on libraries/launcher/org/prismlauncher/utils/JsonParser.java:
```diff
@@ -0,0 +1,425 @@
+/**
+ * Single-file JSON parser to allow for usage in versions without GSON.
+ */
+public final class JsonParser {
```
It would be cleaner and easier to add GSON to the classpath for legacy versions rather than rolling your own. We don't need to reinvent the wheel.

TheKodeToad replied:
I couldn't think of a clean way to add the library depending on the fix being active, and it may conflict with versions used by old mods (I'm unsure on this)?

Even if it's not removed, this is a parser I've tested extensively.

TheKodeToad replied:
this can possibly be addressed at a later date - but it's nearly been a year :sweat_smile:.

craftycodie commented on libraries/launcher/org/prismlauncher/utils/url/NullConnection.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Spoof 404 response from server to avoid unnecessary requests.
+ */
+public final class NullConnection extends HttpURLConnection {
```
I don't think this is necessary. openConnection doesn't actually establish a connection, the name is kind of misleading. Instead, openConnection().connect() would establish a connection.

craftycodie commented on libraries/launcher/org/prismlauncher/utils/url/UrlUtils.java:
```diff
@@ -0,0 +1,104 @@
+
+/**
+ * A utility class for URLs which uses reflection to access hidden methods.
+ * Unfortunately not supported on newer Java versions.
```
I think the solution I described would support newer java versions, it also seems more lightweight.
In any case, to the uninitiated, it's a bit difficult to understand what this class is doing / what it is for, maybe add more comments?

TheKodeToad replied:
In my testing the current method works with newer Java releases (with the flag), while yours seems to not (*even* with the flag). I did want to use yours, but I'll just have to document mine more.

Btw, the comment saying it's not supported is outdated.

craftycodie commented:
> (I don't think I need to keep quoting the previous message lol)
> 
> Yep, that's why I hadn't already. Maybe it is worth getting rid of the Fix interface though since there's probably only going to be one awkward subclass. Separating individual URL fixes probably isn't going to be something users will want, and other fixes will most likely _not_ be able to utilise this interface.

Agreed

TheKodeToad committed:
Try to make some of the suggested changes

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Skin fix -> online fixes

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Remove MIT license from JsonParser

I don't mind, and it's my code.

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad commented:
Would you be okay with me using the slim skin fix? I know it's Christmas day so I don't expect a response for a while. I'm just that kind of crazy person :stuck_out_tongue:!

Edit: really sorry if this is polluting your (and the other three participants') inbox.

TheKodeToad committed:
Slim skin fix - thanks to @craftycodie and @DelofJ
!

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad commented:
Happy new year!

Since it's been a week I thought I'd remind you in case the comment had been lost in other emails or forgotten about.

craftycodie commented:
Go for it :)

TheKodeToad committed:
Move legacy support classes to another jar

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad commented:
Thank you! I thought it was best to ask because I copied it very closely.

TheKodeToad committed:
Finishing touches

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Fix cloaks conflicting with join server endpoint

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Fix oversights

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Merge remote-tracking branch 'upstream/develop' into skinfix

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Fix cmark again :facepalm:

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Fix screwed up ui file

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

DioEgizio commented on libraries/launcher/legacy/org/prismlauncher/legacy/utils/json/JsonParseException.java:
```diff
@@ -0,0 +1,48 @@
+        super(message);
+    }
+
+}
```
```suggestion
}

```

DioEgizio commented on libraries/launcher/legacy/org/prismlauncher/legacy/utils/url/UrlUtils.java:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+
+}
```
```suggestion
}

```

TheKodeToad committed:
Stupidly small skinfixfix as requested by DioEgizio

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Merge branch 'PrismLauncher:develop' into skinfix

Scrumplex commented:
We dropped Snap packaging in #929. You can rebase / merge develop to get rid of the failed task in the CI

TheKodeToad committed:
Merge branch 'PrismLauncher:develop' into skinfix

DioEgizio reviewed (approved):
<img width="966" alt="works" src="https://user-images.githubusercontent.com/83089242/233316704-f33cf805-869f-4609-8cb6-bf66f7da19f0.png">
seeems to work for me

flowln commented on libraries/launcher/legacy/org/prismlauncher/legacy/fix/online/OnlineFixes.java:
```diff
@@ -0,0 +1,81 @@
+
+        if (!UrlUtils.isSupported() || !Base64.isSupported()) {
+            Log.warning("Cannot access the necessary Java internals for skin fix");
+            Log.warning("Turning off legacy skin fix in Settings > Miscellaneous will silence the warnings");
```
```suggestion
            Log.warning("Turning off online fixes in the settings will silence the warnings");
```
what are the odds the placement will change in the future, and no one remembers to update this string :p

flowln commented on libraries/launcher/legacy/org/prismlauncher/legacy/fix/online/OnlineFixes.java:
```diff
@@ -0,0 +1,81 @@
+            URL.setURLStreamHandlerFactory(new OnlineFixes());
+        } catch (Error e) {
+            Log.warning("Cannot apply skin fix: URLStreamHandlerFactory is already set");
+            Log.warning("Turning off legacy skin fix in Settings > Miscellaneous will silence the warnings");
```
here too

flowln reviewed (approved):
LGTM (though its based on my very limited java knowledge :p)

i'll mark it as blocked until the meta PR gets mergd

TheKodeToad commented:
Oops! ty for pointing this out :p

TheKodeToad committed:
thanks flowln! :P

Signed-off-by: Kode <TheKodeToad@proton.me>

TheKodeToad commented:
Surely with any level of Java knowledge you can see my code is garbage though ^_^
Seriously there are some interesting hacks here!

TheKodeToad committed:
Merge branch 'PrismLauncher:develop' into skinfix

TheKodeToad commented:
Seems like `java.protocol.handler.pkgs` is a better solution for compatibility; and i think that's how legacyfix is applied :woman_facepalming:

~~I think improving component support is the way forward?~~ still doesn't seem as good as this tbh

TheKodeToad commented:
Hmm, I don't think this needs to be closed just yet actually.

TheKodeToad committed:
Merge branch 'develop' into skinfix

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad commented:
Seems this is actually more compatible than using `java.protocol.handler.pkgs` or legacyfix :P

TheKodeToad committed:
Use override-based setting for online fixes

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Merge branch 'PrismLauncher:develop' into skinfix

TheKodeToad committed:
Merge remote-tracking branch 'upstream/develop' into skinfix

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Legacy settings override default -> false

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Remove final conflict - kept by mistake

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

ColonelGerdauf reviewed (approved):
Looks like a nice mechanism. The ability to use account or user-defined skins in legacy versions will be nice for people who still want to play the legacy version. Having a skin of "Steve" can be awkward.

Trial97 commented on launcher/Application.cpp:
```diff
@@ -579,6 +579,12 @@ Application::Application(int& argc, char** argv) : QApplication(argc, argv)
         m_settings->registerSetting("IgnoreJavaWizard", false);
 
+        // Mod loader settings
+        m_settings->registerSetting("DisableQuiltBeacon", false);
```
I don't think `QuiltBeacon` is a thing anymore, so consider dropping this.

TheKodeToad replied:
I think I did that by mistake when merging

Trial97 reviewed (approved):
lgtm

TheKodeToad committed:
Drop leftover mod loader settings

I messed up the merge.

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad committed:
Mark setting as optional

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TheKodeToad commented:
Don't merge yet!!

TheKodeToad committed:
Fix small mistake 😭

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

TayouVR reviewed (approved):
my java is a bit rusty, but LGTM

Trial97 commented:
~Added back the blocked label as the meta PR was reverted~

Scrumplex commented:
gg

craftycodie commented:
Nice work @TheKodeToad well done
